src/frontend/src/session.rs

```diff
@@ -645,22 +645,27 @@ impl Session<PgResponseStream> for SessionImpl {
         // false: TEXT
         // true: BINARY
         format: bool,
-    ) -> std::result::Result<PgResponse<PgResponseStream>, BoxedError> {
+    ) -> std::result::Result<(PgResponse<PgResponseStream>, bool), BoxedError> {
         // Parse sql.
         let mut stmts = Parser::parse_sql(sql)
             .inspect_err(|e| tracing::error!("failed to parse sql:\n{}:\n{}", sql, e))?;
         if stmts.is_empty() {
-            return Ok(PgResponse::empty_result(
+            return Ok((PgResponse::empty_result(
                 pgwire::pg_response::StatementType::EMPTY,
-            ));
+            ), false));
         }
         if stmts.len() > 1 {
-            return Ok(PgResponse::empty_result_with_notice(
+            return Ok((PgResponse::empty_result_with_notice(
                 pgwire::pg_response::StatementType::EMPTY,
                 "cannot insert multiple commands into statement".to_string(),
-            ));
+            ), false));
         }
         let stmt = stmts.swap_remove(0);
+        let must_be_query = if let Statement::Insert { returning, .. } = stmt.clone() {
```

Comment:
What's this boolean for?

Reply:
The boolean describes if the output INSERT statement should be like a query, used in src/utils/pgwire/src/pg_protocol.rs line 328

Reply:
additional thoughts: since pg_protocol.rs cannot access the parser, this job has to be done by run_statement

Reply:
The statement type is decided by the frontend. I guess we can add 3 new `StatementType`s like `INSERT_RETURNING`, and make it `is_query` as well?

https://github.com/risingwavelabs/risingwave/blob/241d9c26ebe578bbccd50605a23c939c614b3c17/src/frontend/src/handler/query.rs#L215-L227

https://github.com/risingwavelabs/risingwave/blob/241d9c26ebe578bbccd50605a23c939c614b3c17/src/utils/pgwire/src/pg_response.rs#L131-L139

Reply:
good idea! will work on that